hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -479,6 +510,200 @@ public Hashtable<String, String> getPathStatus(final Path path,
+   * @param srcPath source path
+   * @param dstPath destination path
+   * @param tracingContext object of TracingContext used for the tracing of the
+   * server calls.
+   * @throws AzureBlobFileSystemException exception thrown from the server calls,
+   * or if it is discovered that the copying is failed or aborted.
+   */
+  @org.apache.hadoop.classification.VisibleForTesting
+  void copyBlob(Path srcPath,
+      Path dstPath,
+      TracingContext tracingContext) throws AzureBlobFileSystemException {
+    AbfsRestOperation copyOp = null;
+    try {
+      copyOp = client.copyBlob(srcPath, dstPath,
+          tracingContext);
+    } catch (AbfsRestOperationException ex) {
+      if (ex.getStatusCode() == HttpURLConnection.HTTP_CONFLICT) {
+        final BlobProperty dstBlobProperty = getBlobProperty(dstPath,
+            tracingContext);
+        try {
+          if (dstBlobProperty.getCopySourceUrl() != null &&
+              ("/" + client.getFileSystem() + srcPath.toUri().getPath()).equals(
+                  new URL(dstBlobProperty.getCopySourceUrl()).toURI()
+                      .getPath())) {
+            return;
+          }
+        } catch (URISyntaxException | MalformedURLException e) {
+          throw new RuntimeException(e);
+        }
+      }
+      throw ex;
+    }
+    final String progress = getCopyBlobProgress(copyOp);
+    if (COPY_STATUS_SUCCESS.equals(progress)) {
+      return;
+    }
+    final String copyId = copyOp.getResult().getResponseHeader(X_MS_COPY_ID);
+    while (true) {//TODO: better condition.
+      if (handleCopyInProgress(dstPath, tracingContext, copyId)) {
+        return;
+      }
+      try {
+        Thread.sleep(1000l); //Taken sleep time from AzureNativeFileSystemStore.
+      } catch (Exception e) {
+
+      }
+    }
+  }
+
+  /**
+   * Verifies if the blob copy is success or a failure or still in progress.
+   * @param dstPath path of the destination for the copying
+   * @param tracingContext object of tracingContext used for the tracing of the
+   * server calls.
+   * @param copyId id returned by server on the copy server-call. This id gets
+   * attached to blob and is returned by GetBlobProperties API on the destination.
+   *
+   * @return true if copying is success, false if it is still in progress.
+   * @throws AzureBlobFileSystemException exception returned in making server call
+   * for GetBlobProperties on the path. It can be thrown if the copyStatus is failure
+   * or is aborted.
+   */
+  @org.apache.hadoop.classification.VisibleForTesting
+  boolean handleCopyInProgress(final Path dstPath,
+      final TracingContext tracingContext,
+      final String copyId) throws AzureBlobFileSystemException {
+    BlobProperty blobProperty = getBlobProperty(dstPath,
+        tracingContext);
+    if (blobProperty != null && copyId.equals(blobProperty.getCopyId())) {
+      if (COPY_STATUS_SUCCESS.equals(blobProperty.getCopyStatus())) {
```

Comment:
Should this be case insensitive ?

Reply:
taken. Though not important. Because, spellings for const taken from the api doc: https://learn.microsoft.com/en-us/rest/api/storageservices/get-blob-properties?tabs=azure-ad#response-headers `x-ms-copy-status: <pending \| success \| aborted \| failed>`